KAFKA-15476: Resolve checkstyle cache miss

tylerbertrand opened the pull request:
Resolves cache misses in `checkstyle` tasks due to absolute paths in `configProperties`.

Sets `configDirectory` extension property, which is made available by the `checkstyle` plugin as `${config_loc}` in the `checkstyle` xml files, as shown in the Checkstyle Gradle docs [here](https://docs.gradle.org/current/userguide/checkstyle_plugin.html#sec:checkstyle_built_in_variables). The absolute paths set in `configProperties` are then replaced by relative paths from `configDirectory`. Because the header and suppression config file names are static and only referenced once, these were removed from `configProperties` and the file names are given directly in `checkstyle.xml` 

[Task input comparison showing input differences between builds causing cache miss](https://ge.solutions-team.gradle.com/c/udumk6sv5uwbw/an5e6pclyiblm/task-inputs?cacheability=cacheable,overlapping-outputs,validation-failure)
[Task input comparison with fix applied - no differences between builds](https://ge.solutions-team.gradle.com/c/ojlblvuooowjo/rrubheuf5ex5g/task-inputs?cacheability=cacheable,overlapping-outputs,validation-failure)

### Tests

Manually verified suppressions are working as expected:
* [checkStyleMain FAILS when calling `System.exit()` in `MessageGenerator.java` with that suppression DISABLED](https://ge.solutions-team.gradle.com/s/34amuqemmyili/failure#1)
* [checkStyleMain PASSES when calling `System.exit()` in `MessageGenerator.java` with that suppression ENABLED](https://ge.solutions-team.gradle.com/s/s62zwn5gfsu54/failure)

Manually verified header check is working as expected:
* [checkStyleMain FAILS when java file is missing license header](https://ge.solutions-team.gradle.com/s/rjjqrqrmam6oq/failure#1)

tylerbertrand committed:
Resolve checkstyle cache miss

divijvaidya commented on build.gradle:
```diff
@@ -685,6 +685,7 @@ subprojects {
   }
 
   checkstyle {
+    configDirectory = rootProject.layout.projectDirectory.dir("checkstyle")
```
is this rootDir.dir("checkstyle")? or `"File(rootDir, "checkstyle")"`?

My point is, could we make it relative to rootDir here?

tylerbertrand replied:
These all point to the same directory, though `rootProject.layout.projectDirectory.dir("checkstyle")` uses lazy properties, which are generally preferred.

divijvaidya replied:
ok, that makes sense

divijvaidya commented on build.gradle:
```diff
@@ -685,6 +685,7 @@ subprojects {
 
   checkstyle {
+    configDirectory = rootProject.layout.projectDirectory.dir("checkstyle")
     configFile = new File(rootDir, "checkstyle/checkstyle.xml")
```
we can perhaps change the file relative to configDirectory now instead of working relative to root directory.

tylerbertrand replied:
Good point. It turns out `${configDirectory}/checkstyle.xml` is [the default location](https://docs.gradle.org/current/dsl/org.gradle.api.plugins.quality.CheckstyleExtension.html#org.gradle.api.plugins.quality.CheckstyleExtension:configFile), so there is actually no need to specify `configFile` in this case. I pushed a commit to remove that line.

divijvaidya reviewed (commented):
Thank you for the change. Suggested some minor changes, otherwise looks good to me.

tylerbertrand committed:
Use default checkstyle configFile location

By default, the checkstyle plugin looks for the configFile at ${configDirectory}/checkstyle.xml, so there is no need to set the configFile location manually

divijvaidya commented:
Test failures are unrelated

```
[Build / JDK 17 and Scala 2.13 / kafka.zk.ZkMigrationIntegrationTest.[1] Type=ZK, Name=testDualWriteQuotaAndScram, MetadataVersion=3.5-IV2, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/kafka.zk/ZkMigrationIntegrationTest/Build___JDK_17_and_Scala_2_13____1__Type_ZK__Name_testDualWriteQuotaAndScram__MetadataVersion_3_5_IV2__Security_PLAINTEXT/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_11_and_Scala_2_13___testBalancePartitionLeaders__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_11_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManagerTest.testRemoteLogSizeCalculationWithSegmentsOfTheSameEpoch()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/org.apache.kafka.server.log.remote.metadata.storage/TopicBasedRemoteLogMetadataManagerTest/Build___JDK_8_and_Scala_2_12___testRemoteLogSizeCalculationWithSegmentsOfTheSameEpoch__/)
[Build / JDK 20 and Scala 2.13 / kafka.api.ConsumerBounceTest.testConsumptionWithBrokerFailures()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/kafka.api/ConsumerBounceTest/Build___JDK_20_and_Scala_2_13___testConsumptionWithBrokerFailures__/)
[Build / JDK 20 and Scala 2.13 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testCreateTokenForOtherUserFails(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/kafka.api/DelegationTokenEndToEndAuthorizationWithOwnerTest/Build___JDK_20_and_Scala_2_13___testCreateTokenForOtherUserFails_String__quorum_kraft/)
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_20_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
[Build / JDK 17 and Scala 2.13 / kafka.zk.ZkMigrationIntegrationTest.[1] Type=ZK, Name=testDualWriteQuotaAndScram, MetadataVersion=3.5-IV2, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/kafka.zk/ZkMigrationIntegrationTest/Build___JDK_17_and_Scala_2_13____1__Type_ZK__Name_testDualWriteQuotaAndScram__MetadataVersion_3_5_IV2__Security_PLAINTEXT/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_11_and_Scala_2_13___testBalancePartitionLeaders__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_11_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManagerTest.testRemoteLogSizeCalculationWithSegmentsOfTheSameEpoch()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/org.apache.kafka.server.log.remote.metadata.storage/TopicBasedRemoteLogMetadataManagerTest/Build___JDK_8_and_Scala_2_12___testRemoteLogSizeCalculationWithSegmentsOfTheSameEpoch__/)
[Build / JDK 20 and Scala 2.13 / kafka.api.ConsumerBounceTest.testConsumptionWithBrokerFailures()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/kafka.api/ConsumerBounceTest/Build___JDK_20_and_Scala_2_13___testConsumptionWithBrokerFailures__/)
[Build / JDK 20 and Scala 2.13 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testCreateTokenForOtherUserFails(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/kafka.api/DelegationTokenEndToEndAuthorizationWithOwnerTest/Build___JDK_20_and_Scala_2_13___testCreateTokenForOtherUserFails_String__quorum_kraft/)
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14344/2/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_20_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
```

divijvaidya commented:
@tylerbertrand congratulations on your first contributions to this project! 

I performed some housekeeping for you in this case such as creating a [JIRA](https://issues.apache.org/jira/browse/KAFKA-15476), tagging it with PR title etc. For future contributions, I would encourage you to read the contribution guide - https://kafka.apache.org/contributing.html
